controllers/k6_initialize.go

```diff
@@ -0,0 +1,197 @@
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/grafana/k6-operator/api/v1alpha1"
+	"github.com/grafana/k6-operator/pkg/cloud"
+	"github.com/grafana/k6-operator/pkg/resources/jobs"
+	"github.com/grafana/k6-operator/pkg/types"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// k6 Cloud related vars
+// Right now operator works with one test at a time so these should be safe.
+var (
+	testRunId string
+	token     string
+)
+
+// InitializeJobs creates jobs that will run initial checks for distributed test if any are necessary
+func InitializeJobs(ctx context.Context, log logr.Logger, k6 *v1alpha1.K6, r *K6Reconciler) (res ctrl.Result, err error) {
+	log.Info("Initialize test")
+
+	k6.Status.Stage = "initialization"
+	if err = r.Client.Status().Update(ctx, k6); err != nil {
+		log.Error(err, "Could not update status of custom resource")
+		return
+	}
+
+	if cli := types.ParseCLI(&k6.Spec); cli.HasCloudOut {
+		var (
+			secrets    corev1.SecretList
+			secretOpts = &client.ListOptions{
+				// TODO: find out a better way to get namespace here
+				Namespace: "k6-operator-system",
+				LabelSelector: labels.SelectorFromSet(map[string]string{
+					"k6cloud": "token",
+				}),
+			}
+		)
+		if err := r.List(ctx, &secrets, secretOpts); err != nil {
+			log.Error(err, "Failed to load k6 Cloud token")
+			return res, err
+		}
+
+		if len(secrets.Items) < 1 {
+			err := fmt.Errorf("There are no secrets to hold k6 Cloud token")
+			log.Error(err, err.Error())
+			return res, err
+		}
+
+		if t, ok := secrets.Items[0].Data["token"]; !ok {
+			err := fmt.Errorf("The secret doesn't have a field token for k6 Cloud")
+			log.Error(err, err.Error())
+			return res, err
+		} else {
+			token = string(t)
+		}
```

Comment:
Can we consider trying to load the token from other sources as well? For instance we typically don't use k8s secrets, instead we use Vault with a side car to mount a secrets volume for the container. In our case, with Vault, we were hoping to use the file configuration (https://k6.io/docs/cloud/integrations/token/#authentication-with-a-config-file). Is it possible to look for the other methods of providing a token as documented at the link above, then also look for k8s secret? While the env var approach is certainly not as secure, it does come in handy for testing...however, at the very least file config support would be appreciated. :)

Reply:
Hi @brian-groux-hs! Sorry for such a delay in response.
Yes, I definitely would like to have alternative options to load secrets here. I wasn't sure what people might use so you just gave me an answer to that, thank you :smile: I'm not sure at the moment if operator should try to load everything one by one as you describe or be a separate configuration option though...
Either way, I won't be able to add it right now but noted in TODO. I'll probably create a separate issue for this.

Reply:
Related issue: https://github.com/grafana/k6-operator/issues/103